### Zadanie 1

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv('../Lab2/phpMYEkMl.csv', index_col='id')
df = df.replace('?', np.nan)

new_column_name = []
for col in df.columns:
    new_column_name.append(col[1:-1])
df.columns = new_column_name

df['CabinReduced'] = df['cabin'].astype(str).str[0]

df.head(20)

pclass  survived                                              name  \
id                                                                       
1        1         1                      Allen,Miss. Elisabeth Walton   
2        1         1                     Allison,Master. Hudson Trevor   
3        1         0                       Allison,Miss. Helen Loraine   
4        1         0               Allison,Mr. Hudson Joshua Creighton   
5        1         0    Allison,Mrs. Hudson J C (Bessie Waldo Daniels)   
6        1         1                                Anderson,Mr. Harry   
7        1         1                  Andrews,Miss. Kornelia Theodosia   
8        1         0                             Andrews,Mr. Thomas Jr   
9        1         1      Appleton,Mrs. Edward Dale (Charlotte Lamson)   
10       1         0                            Artagaveytia,Mr. Ramon   
11       1         0                             Astor,Col. John Jacob   
12       1         1  Astor,Mrs. John Jacob (Madeleine Talmadge Force)   
13       1         1                      Aubart,Mme. Leontine Pauline   
14       1         1                       Barber,Miss. Ellen 'Nellie'   
15       1         1               Barkworth,Mr. Algernon Henry Wilson   
16       1         0                                Baumann,Mr. John D   
17       1         0                           Baxter,Mr. Quigg Edmond   
18       1         1    Baxter,Mrs. James (Helene DeLaudeniere Chaput)   
19       1         1                              Bazzani,Miss. Albina   
20       1         0                               Beattie,Mr. Thomson   

       sex     age  sibsp  parch    ticket      fare    cabin embarked boat  \
id                                                                            
1   female      29      0      0     24160  211.3375       B5        S    2   
2     male  0.9167      1      2    113781    151.55  C22 C26        S   11   
3   female       2      1      2    113781    151.55  C22 C26        S  NaN   
4     male      30      1      2    113781    151.55  C22 C26        S  NaN   
5   female      25      1      2    113781    151.55  C22 C26        S  NaN   
6     male      48      0      0     19952     26.55      E12        S    3   
7   female      63      1      0     13502   77.9583       D7        S   10   
8     male      39      0      0    112050         0      A36        S  NaN   
9   female      53      2      0     11769   51.4792     C101        S    D   
10    male      71      0      0  PC 17609   49.5042      NaN        C  NaN   
11    male      47      1      0  PC 17757   227.525  C62 C64        C  NaN   
12  female      18      1      0  PC 17757   227.525  C62 C64        C    4   
13  female      24      0      0  PC 17477      69.3      B35        C    9   
14  female      26      0      0     19877     78.85      NaN        S    6   
15    male      80      0      0     27042        30      A23        S    B   
16    male     NaN      0      0  PC 17318    25.925      NaN        S  NaN   
17    male      24      0      1  PC 17558  247.5208  B58 B60        C  NaN   
18  female      50      0      1  PC 17558  247.5208  B58 B60        C    6   
19  female      32      0      0     11813   76.2917      D15        C    8   
20    male      36      0      0     13050   75.2417       C6        C    A   

   body                      home.dest CabinReduced  
id                                                   
1   NaN                    St Louis,MO            B  
2   NaN  Montreal,PQ / Chesterville,ON            C  
3   NaN  Montreal,PQ / Chesterville,ON            C  
4   135  Montreal,PQ / Chesterville,ON            C  
5   NaN  Montreal,PQ / Chesterville,ON            C  
6   NaN                    New York,NY            E  
7   NaN                      Hudson,NY            D  
8   NaN                     Belfast,NI            A  
9   NaN              Bayside,Queens,NY            C  
10   22             Montevideo,Uruguay    

### Zadanie 2
Funkcja train_test_split, umożliwia podział zestawu danych na zbiór treningowy i testowy.
- Pozwala na losowe podzielenie danych
- Możemy określić rozmiar zbioru testowego
- Można utworzyć zbiór treningowy i testowy

### Zadanie 3
Stwórz zmienną do której zapiszesz listę z nazwami trzech kolumn –
kabiny, zredukowane kabiny oraz płeć. Nazwij ją col_name

In [2]:
col_name = ['CabinReduced', 'cabin', 'sex']
# df[['CabinReduced', 'cabin', 'sex']]

### Zadanie 4
Jako zmienną niezależną ustaw dane składające się z kolumn o nazwach
zapisanych w col_name. Jako zmienną zależną ustaw kolumnę mówiącą o
tym czy ktoś przeżył czy nie. Ustaw rozmiar zbioru testowego na 20 lub
30% całości. Parametr random_state ustaw jako 0.
Wyświetl wymiary zbiorów X_train, X_test, y_train, y_test używając
.shape() i skomentuj wyniki.

In [3]:
from sklearn.model_selection import train_test_split

zmienna_niezalezna = df[col_name]
zmienna_zalezna = df['survived']

X_train, X_test, y_train, y_test = train_test_split(zmienna_niezalezna, zmienna_zalezna, test_size=0.2, random_state=0)
print(f'Wymiar X_train: {X_train.shape}')
print(f'Wymiar X_test: {X_test.shape}')
print(f'Wymiar y_train: {y_train.shape}')
print(f'Wymiar y_test: {y_test.shape}')

Wymiar X_train: (1047, 3)
Wymiar X_test: (262, 3)
Wymiar y_train: (1047,)
Wymiar y_test: (262,)


Komentarz wyników
- parametr random_state ustawiony na 0, zapewni powtarzalność wyników podczas kolejnych uruchomień programu,
- zbiór treningowy zawiera 1047 wierszy oraz 3 kolumny
- zbiór testowy zawiera 262 wierszy oraz 3 kolumny
- zbiór treningowy zmiennej zależnej zawiera 1047 wartości
- zbiór testowy zmiennej zależnej zawiera 262 wartości

### Zadanie 5
Zbadaj w zależności od kardynalności danej zmiennej czy rozkład etykiet
dla poszczególnych cech w zbiorach testowych i treningowych jest
równomierny. 

In [4]:
for col in col_name:
    temp = X_train[col].unique()
    unique_test = [x for x in temp if x not in X_test[col].unique()]
    print(f'Kolumna {col}:')
    print(f'Liczba unikalnych wartości w zbiorze treningowym: {len(temp)}')
    print(f'Liczba unikalnych wartości w zbiorze treningowym, które nie występują w zbiorze testowym: {len(unique_test)}', end='\n\n')


Kolumna CabinReduced:
Liczba unikalnych wartości w zbiorze treningowym: 9
Liczba unikalnych wartości w zbiorze treningowym, które nie występują w zbiorze testowym: 1

Kolumna cabin:
Liczba unikalnych wartości w zbiorze treningowym: 164
Liczba unikalnych wartości w zbiorze treningowym, które nie występują w zbiorze testowym: 139

Kolumna sex:
Liczba unikalnych wartości w zbiorze treningowym: 2
Liczba unikalnych wartości w zbiorze treningowym, które nie występują w zbiorze testowym: 0



### Zadanie 6
Wykonaj kodowanie zmiennych kategorycznych do zmiennych
liczbowych. Wykorzystaj pętlę for i metodę enumerate()

In [5]:
def categorical_coding(df, col):
    categorical_data = df[col].unique()
    label_dict = {}
    for i, category in enumerate(categorical_data):
        label_dict[category] = i
    return label_dict

CabinReduced_map_train = categorical_coding(X_train, col_name[0])
cabin_map_train = categorical_coding(X_train, col_name[1])
sex_map_train = categorical_coding(X_train, col_name[2])


CabinReduced_map_test = categorical_coding(X_test, col_name[0])
cabin_map_test = categorical_coding(X_test, col_name[1])
sex_map_test = categorical_coding(X_test, col_name[2])

print('CabinReduced zbiór X_train:')
print(CabinReduced_map_train)
print('\ncabin zbiór X_train:')
print(cabin_map_train)
print('\nsex zbiór X_train:')
print(sex_map_train)
print('\nCabinReduced zbiór X_test:')
print(CabinReduced_map_test)
print('\ncabin zbiór X_test:')
print(cabin_map_test)
print('\nsex zbiór X_test:')
print(sex_map_test)

CabinReduced zbiór X_train:
{'n': 0, 'E': 1, 'F': 2, 'A': 3, 'C': 4, 'D': 5, 'B': 6, 'T': 7, 'G': 8}

cabin zbiór X_train:
{nan: 0, 'E40': 1, 'F G73': 2, 'A19': 3, 'C32': 4, 'D': 5, 'B79': 6, 'C62 C64': 7, 'E8': 8, 'D28': 9, 'C45': 10, 'C46': 11, 'B22': 12, 'B57 B59 B63 B66': 13, 'C23 C25 C27': 14, 'E44': 15, 'B45': 16, 'B39': 17, 'B58 B60': 18, 'F2': 19, 'C47': 20, 'B101': 21, 'D21': 22, 'A7': 23, 'E52': 24, 'F38': 25, 'E36': 26, 'C68': 27, 'E24': 28, 'C22 C26': 29, 'D38': 30, 'B50': 31, 'A24': 32, 'C111': 33, 'F': 34, 'C6': 35, 'C87': 36, 'C93': 37, 'D36': 38, 'C125': 39, 'B35': 40, 'T': 41, 'B73': 42, 'A26': 43, 'A18': 44, 'B96 B98': 45, 'G6': 46, 'C78': 47, 'C101': 48, 'D9': 49, 'D33': 50, 'C128': 51, 'E50': 52, 'B26': 53, 'B69': 54, 'E121': 55, 'C123': 56, 'B94': 57, 'A34': 58, 'C39': 59, 'D43': 60, 'E31': 61, 'B5': 62, 'D17': 63, 'F33': 64, 'D7': 65, 'A21': 66, 'D34': 67, 'A29': 68, 'D35': 69, 'A11': 70, 'B51 B53 B55': 71, 'D46': 72, 'E60': 73, 'C30': 74, 'D26': 75, 'E68': 76, 'A

In [6]:
X_train['CabinReduced_map'] = X_train['CabinReduced'].map(CabinReduced_map_train)
X_train['cabin_map'] = X_train['cabin'].map(cabin_map_train)
X_train['sex_map'] = X_train['sex'].map(sex_map_train)
X_test['CabinReduced_map'] = X_test['CabinReduced'].map(CabinReduced_map_test)
X_test['cabin_map'] = X_test['cabin'].map(cabin_map_test)
X_test['sex_map'] = X_test['sex'].map(sex_map_test)

print('Zbior treningowy')
print(X_train.head(20))

print('\nZbior testowy')
print(X_test.head(20))


Zbior treningowy
     CabinReduced  cabin     sex  CabinReduced_map  cabin_map  sex_map
id                                                                    
1119            n    NaN    male                 0          0        0
45              E    E40  female                 1          1        1
1073            n    NaN    male                 0          0        0
1131            n    NaN  female                 0          0        1
575             n    NaN    male                 0          0        0
1218            F  F G73    male                 2          2        0
501             n    NaN    male                 0          0        0
959             n    NaN  female                 0          0        1
270             A    A19    male                 3          3        0
323             C    C32  female                 4          4        1
704             n    NaN    male                 0          0        0
669             n    NaN    male                 0          

### Zadanie 8
Sprawdź liczbę brakujących wartości w zmodyfikowanych zbiorach.
Zapisz wyniki i skomentuj. 

In [7]:
test_nan_CabinReduced_map = X_test['CabinReduced_map'].isna().sum()
test_nan_cabin_map = X_test['cabin_map'].isna().sum()
test_nan_sex_map = X_test['sex_map'].isna().sum()

train_nan_CabinReduced_map = X_train['CabinReduced_map'].isna().sum()
train_nan_cabin_map = X_train['cabin_map'].isna().sum()
train_nan_sex_map = X_train['sex_map'].isna().sum()

print(f'W zbiorze testowym kolumna CabinReduced_map ma {test_nan_CabinReduced_map} brakujacych wartosci.')
print(f'W zbiorze testowym kolumna cabin_map ma {test_nan_cabin_map} brakujacych wartosci.')
print(f'W zbiorze testowym kolumna sex_map ma {test_nan_sex_map} brakujacych wartosci.')

test_nan = X_test[col_name].isna().sum()
print('\nZbiór testowy dla pozostałych klumn - suma brakujących wartosci w kolumnie')
print(test_nan, end='\n\n')

print(f'W zbiorze treningowym kolumna CabinReduced_map ma {train_nan_CabinReduced_map} brakujacych wartosci.')
print(f'W zbiorze treningowym kolumna cabin_map ma {train_nan_cabin_map} brakujacych wartosci.')
print(f'W zbiorze treningowym kolumna sex_map ma {train_nan_sex_map} brakujacych wartosci.')

train_nan = X_train[col_name].isna().sum()
print('\nZbiór treningowy dla pozostałych klumn - suma brakujących wartosci w kolumnie')
print(train_nan, end='\n\n')


W zbiorze testowym kolumna CabinReduced_map ma 0 brakujacych wartosci.
W zbiorze testowym kolumna cabin_map ma 0 brakujacych wartosci.
W zbiorze testowym kolumna sex_map ma 0 brakujacych wartosci.

Zbiór testowy dla pozostałych klumn - suma brakujących wartosci w kolumnie
CabinReduced      0
cabin           211
sex               0
dtype: int64

W zbiorze treningowym kolumna CabinReduced_map ma 0 brakujacych wartosci.
W zbiorze treningowym kolumna cabin_map ma 0 brakujacych wartosci.
W zbiorze treningowym kolumna sex_map ma 0 brakujacych wartosci.

Zbiór treningowy dla pozostałych klumn - suma brakujących wartosci w kolumnie
CabinReduced      0
cabin           803
sex               0
dtype: int64



W zbiorze testowym oraz treningowym nie ma brakujących wartości w kolumnach, do których zostały dokonane kodowania zmiennych kategorycznych. <br>W zbiorze testowym w kolumnie cabin są 211 brakujących wartości.<br>W zbiorze treningowym w kolumnie cabin są 803 brakujących wartości.<br><br>To sugeruje, że ta kolumna może być trudna do zastosowania w modelowaniu predykcyjnym.

### Zadanie 9
Zastąp brakujące wartości liczbą 0. Czy jest to najlepsze wyjście?

Odpowiedź:
Można zastosować
- uzupełnianie średnią lub medianą,
- usunięcie rekordów z brakującymi wartościami, 
- uzupełnienie brakujących wartości na podstawie innych danych.

In [8]:
X_test['cabin'] = X_test['cabin'].fillna(0)
X_train['cabin'] = X_train['cabin'].fillna(0)

### Zadanie 10
Porównaj ile unikalnych wartości jest w zbiorze treningowym, a ile w
zbiorze testowym (funkcja len).
Jaka jest różnica pomiędzy liczbą etykiet przed i po redukcji oraz
mapowaniu?
Czy cały proces, który został do tej pory wykonany może mieć wpływ na
końcowy wynik predykcji i jakość modelu?

In [9]:
for col in X_train.columns:
    print(f'Kolumna {col}:')
    print('Liczba unikalnych wartości w zbiorze treningowym:', len(X_train[col].unique()))
    print('Liczba unikalnych wartości w zbiorze testowym:', len(X_test[col].unique()), end='\n\n')

Kolumna CabinReduced:
Liczba unikalnych wartości w zbiorze treningowym: 9
Liczba unikalnych wartości w zbiorze testowym: 8

Kolumna cabin:
Liczba unikalnych wartości w zbiorze treningowym: 164
Liczba unikalnych wartości w zbiorze testowym: 49

Kolumna sex:
Liczba unikalnych wartości w zbiorze treningowym: 2
Liczba unikalnych wartości w zbiorze testowym: 2

Kolumna CabinReduced_map:
Liczba unikalnych wartości w zbiorze treningowym: 9
Liczba unikalnych wartości w zbiorze testowym: 8

Kolumna cabin_map:
Liczba unikalnych wartości w zbiorze treningowym: 164
Liczba unikalnych wartości w zbiorze testowym: 49

Kolumna sex_map:
Liczba unikalnych wartości w zbiorze treningowym: 2
Liczba unikalnych wartości w zbiorze testowym: 2



Różnica pomiędzy liczbą etykiet przed i po redukcji oraz mapowaniu zachodzi tylko w kolumnie cabin,<br>Przed:<br>
  Liczba unikalnych wartości w zbiorze testowym: 30<br>
Po:<br>
  Liczba unikalnych wartości w zbiorze testowym: 49<br><br><br>
Wpływ na jakość predykcji może mieć fakt, że część klas lub wartości została po prostu usunięta lub zmapowana na inną wartość.<br>
Redukcja liczby etykiet może pomóc w uproszczeniu modelu i zwiększeniu jego wydajności. Ponieważ liczba unikalnych wartości zwiększyła się.